sigstore/_sign.py

```diff
@@ -0,0 +1,114 @@
+import base64
+import hashlib
+
+import jwt
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import ec
+
+from sigstore._internal.fulcio import (
+    FulcioCertificateSigningRequest,
+    FulcioClient,
+)
+from sigstore._internal.rekor import RekorClient
+from sigstore._internal.oidc import Identity
+
+# from cryptography.x509 import load_pem_x509_certificate
+
+
+REKOR_URL = "https://rekor.sigstore.dev"
+REKOR_API_HEADERS = {"Content-Type": "application/json", "Accept": "application/json"}
+
+
+def _no_output(*a, **kw):
+    pass
+
+
+def sign(file_, identity_token, output=_no_output):
+    """Public API for signing blobs"""
+
+    output(f"Using payload from: {file_.name}")
+    artifact_contents = file_.read().encode()
+    sha256_artifact_hash = hashlib.sha256(artifact_contents).hexdigest()
+
+    output("Generating ephemeral keys...")
+    # Generate the ephemeral private key
+    # private_key = Ed25519PrivateKey.generate()
+    # public_key = private_key.public_key()
+    private_key = ec.generate_private_key(ec.SECP384R1())
+    public_key = private_key.public_key()
+
+    output("Retrieving signed certificate...")
+    # fulcio = FulcioClient("http://localhost:5555")
+    fulcio = FulcioClient()
+
+    oidc_identity = Identity(identity_token)
+
+    # Build an X.509 Certificiate Signing Request - not currently supported
+    """
+    builder = (
```

Comment:
Is this an API that is proposed for `cryptography`?

Reply:
This is supported for `cryptography` (https://cryptography.io/en/latest/x509/reference/#cryptography.x509.CertificateSigningRequestBuilder) but sending a CSR to Fulcio is what's currently unsupported: https://github.com/sigstore/fulcio/issues/503